Fix the IR gen for C++ method calls and refactor around CGFunctionInfo

hjyamauchi opened the pull request:
This case was missed by https://github.com/swiftlang/swift/pull/76159.

This is a partial fix for https://github.com/swiftlang/swift/issues/74866

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@swift-ci please test macOS platform

hjyamauchi commented:
@swift-ci please test macOS platform

hjyamauchi commented:
@swift-ci please test macOS platform

hjyamauchi commented:
@swift-ci please test macOS platform

hjyamauchi commented:
@swift-ci please test macOS platform

rjmccall reviewed (commented):
Why is this going through this path instead of using the proper call-lowering path?

hjyamauchi commented:
Do you mean why all attributes aren't added in `SignatureExpansion::expandExternalSignatureTypes`?
I wonder that, too. I don't know this code enough to know off hand.

Would looking at these commits, which seem to have added or touched this code, help to clarify?
https://github.com/swiftlang/swift/commit/0e31acb18c109610898a282e69a53ca68c8550dc
https://github.com/swiftlang/swift/commit/b3f302b96b5e099d26362804e750e0a926be6361

rjmccall commented:
@egorzhdan Can you help out here?  Why does the `FunctionPointer` not just have the right attributes for these calls?

egorzhdan commented:
I'm afraid I don't know the reason for the missing attribute. Perhaps this sret/noreg logic should be moved to `SignatureExpansion::expandExternalSignatureTypes`?

@ahatanaka WDYT?

hjyamauchi commented:
@rjmccall WDYT? I think I will see what I can figure out. If there are some hints or places to look at, it'd be great if you can point me to them.

hjyamauchi commented:
Rebased. @swift-ci please test

compnerd commented:
@ahatanaka ping

hjyamauchi commented:
@swift-ci please test

ahatanaka commented:
It looks like it's possible to add `sret` and `inreg` in `SignatureExpansion::expandExternalSignatureTypes`, but special care must be taken to handle noreturn functions correctly as SILGen changes the return type to `Never`.

For example, this is the SIL for the call to `Large largeStructNoReturn()  __attribute__((noreturn))` in `testIndirectReturnNoReturn` in `test/IRGen/noreturn.swift`.

```
  %0 = function_ref @largeStructNoReturn : $@convention(c) () -> Never // user: %1
  %1 = apply %0() : $@convention(c) () -> Never
```

hjyamauchi commented:
@ahatanaka So it sounds like that was the reason to add `sreg` in `IRBuilder::CreateCallOrInvoke`?

@rjmccall What do you think?

rjmccall commented:
SILGen should absolutely not be producing functions with a return type of `Never` just because they're `noreturn`.  I guess it would work if we reliably had a Clang type available on the `SILFunctionType` in IRGen, but otherwise, no; we need the actual Clang type in order to emit the call correctly.  As a workaround, trying to retroactively paper over the problem will often work because most `noreturn` calls will be direct, but it'll blow up if they're ever indirect, and Clang does support `noreturn` function pointers, ObjC methods, `virtual` methods, and so on.

ahatanaka commented:
`ClangImporter::Implementation::importFunctionParamsAndReturnType` is setting the result type:

```
  if (clangDecl->isNoReturn())
    swiftResultTy = SwiftContext.getNeverType();
```

rjmccall commented:
Okay.  That's not an unreasonable decision for the surface language, but it's incorrect for SIL, at least without a guarantee of carrying a C function type in the SIL function type.  We need to know the actual C calling convention of the function in order to call it correctly, and in general that includes knowing its true declared return type.  In theory, the calling conventions for `void (int)` and `float (int)` could be completely different.

Now, on most platforms, we wouldn't actually need to know the *exact* return type.  The call doesn't return, so anything in the ABI that only matters on the return path is irrelevant, which means that e.g. we can absolutely call a `float (int)` as if it were a `void (int)`. In every ABI for machine-code platforms that I know of, the return type only affects the non-return path when the return value is returned indirectly, so maybe we could represent that in SIL function types, e.g. as `@out Never` (which is indirect) vs. `Never` (which is direct).

However:
1. That doesn't really work with how we lower foreign calls with clang, because we'd need to use a return type that forced Clang to return indirectly, which we don't have a way to do; we just assume clang will generate an indirect parameter/return convention for a parameter/return type that's `@in` / `@out` in SIL.
2. That's an unportable assumption because, as noted, the C ABI is absolutely allowed to use completely different calling conventions for different function signatures.
3. I believe WebAssembly does a runtime check that the function signature of a call exactly matches the signature of the callee, at least up to pointers.  So IIUC calling a `float (int)` as a `void (int)` in WASM does not work, even if the function never returns.  Therefore, we really do need the exact C return type in order to do the call correctly.

The cleanest solution would simply be to allow the SIL function type to lie arbitrarily about the return type as long as it preserves the C function type.  IRGen would then just have to recognize when a `Never` return type actually matches a non-trivial return convention in the `CGFunctionInfo` and e.g. fake an indirect argument as appropriate.  But that would require us to actually preserve C function types in SIL function types.

Unfortunately, the alternatives are much more invasive.  We could recognize a C function with a supposed Swift return type of `Never` and try to restore the original return type in that function's SIL type, but for that SIL type, we'd actually need a Swift type that corresponds to that original C return type, which is not something that SIL type lowering knows how to do in general.  And using a SIL return type other than `Never` would actually mess up our SIL analyses, because we'd lose the information that the call didn't return.  And the conservative solution of just not importing non-`void` `noreturn` functions as returning `Never` would break source compatibility for similar reasons.

Maybe there's some viable path for preserving the Clang function type in the SIL function type of just these functions, so that we don't have to tackle the general problem just to fix this bug?

eeckstein commented:
Don't we store the clang type in SILFunctionType already? https://github.com/swiftlang/swift/blob/0d747d4d2b1d6eac0065ea3fc0f3f32d5cf806c1/include/swift/AST/Types.h#L5395

rjmccall commented:
I believe that, just like unlowered function types, it's not reliably there unless you pass the clang-function-types flag.

hjyamauchi commented:
@rjmccall What do you think of getting this PR merged and moving this discussion to a github issue? It aims to address an orthogonal issue of adding a missing `inreg` to the `sret` which seems needed regardless even if we decided to add `sret` in `expandExternalSignatureTypes` instead or elsewhere.

This is one of the PRs that are needed to resurrect the Windows ARM64 Swift compiler and while I don't fully follow the discussion here but it sounds like there's more general issue here about noreturn functions and preserving the Clang function type in the SIL function type.

CC: @ahatanaka  @eeckstein @compnerd

rjmccall commented:
In my experience, it's important to resist solving problems the wrong way in a compiler because the hacks (and bugs) just keep piling up.  In this case, there are a large number of IR attributes that are necessary on various platforms in order to generate calls correctly.  We can already see that this little hack of copying attributes from the callee has been slowly accreting support for more and more attributes.  Each of those accretions presumably represents a separate, slowly-debugged miscompile, each potentially paired with a second and still-latent miscompile when the call actually goes through a function pointer.  There is a central place in IRGen that adds attributes to C-convention calls and declarations based on their type.  Fixing bugs there, and making sure that it's used properly, is how we avoid having to fix bugs ten different ways for ten slightly-different inputs.

eeckstein commented:
@rjmccall In general I agree with what you said. But @hjyamauchi is probably not the right person to do this kind of (I assume) rather big rewrite. We need to get this unblocked. And what @hjyamauchi suggested is the right way to do so. After filing a github issue, whoever owns - and fully understands - the code can do a proper fix as a follow-up.
So, if there isn't any strong objection (e.g. this change would introduce another problem), I would suggest to merge this PR and do the proper fix in a follow-up.

rjmccall commented:
It's not clear to me whether @ahatanaka's observation about `Never` substitution is actually a blocker for fixing this properly in signature expansion.  What happens if we just try to fix it there?

rjmccall commented:
We don't have the exact C function type, but we deconstruct the `SILFunctionType` to create a `CGFunctionInfo`, which should be good enough to recognize that we need to apply `inreg`.  The problem you identified with `Never` is serious but is an existing miscompile that we at least wouldn't be regressing.

ahatanaka commented:
Ah I see. If we don't care about noreturn functions, that should work.

rjmccall commented:
Yeah, I think we clearly need a deeper solution for `noreturn` functions, with the most elegant option being to propagate down the Clang function type.

hjyamauchi commented:
@rjmccall @ahatanaka 

I examined the test case (included in this PR) and it seems that

- The function `getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ` in question doesn't look indirect-return or have an indirect return parameter at the time of `expandExternalSignatureTypes` hence for example it doesn't take the path into the `returnInfo.isIndirect()` block ([here](https://github.com/swiftlang/swift/blob/5285c22ef377befdf4d4cd0126d52dbc2555e7ec/lib/IRGen/GenCall.cpp#L1586)) and no `sret` (nor `inreg`) was added.

- After `expandExternalSignatureTypes`, it's made indirect-return (an indirect return parameter is added) in `SyncCallEmission::setArgs` ([here](https://github.com/swiftlang/swift/blob/5285c22ef377befdf4d4cd0126d52dbc2555e7ec/lib/IRGen/GenCall.cpp#L2563)) in the `passIndirectResults` lambda because`fnConv.getNumIndirectSILResults()` > 0.

- Later in `CreateCallOrInvoke`, when it compares the args to the callee, it finds that the parameter is missing `sret` (and `inreg`) and adds them. This is where this PR currently adds `inreg` and we are discussing whether we could instead do that in `expandExternalSignatureTypes`.

I did some printf debugging around `CGFunctionInfo` in `expandExternalSignatureTypes` but it seemed that `CGFunctionInfo` doesn't have enough info indicating that it is an indirect return or we could add an `sret` there. This is what I got printed:

```
expandExternalSignatureTypes on
(sil_function_type type="@convention(cxx_method) (@in_guaranteed BridgedFunction) -> OptionalBridgedBasicBlock"
  (input=struct_type decl="InRegSRet.(file).BridgedFunction")
  (result=struct_type decl="InRegSRet.(file).OptionalBridgedBasicBlock")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
FnType->getRepresentation() 12
SILResultTy $OptionalBridgedBasicBlock
clangResultTy RecordType 0x59a82d0 'struct OptionalBridgedBasicBlock' imported
`-CXXRecord 0xfe04a8 'OptionalBridgedBasicBlock'
returnInfo.isIndirect 0
getSILFuncConventions().getNumIndirectSILResults() 0
getSILFuncConventions().getNumIndirectSILResults() > 0 0
returnInfo.getKind() 4
returnInfo.isInAlloca() 0
CreateCallOrInvoke: found a missing inreg
declare dso_local void @"?getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ"(ptr noundef nonnull align 1 dereferenceable(1) %0, ptr inreg sret(%struct.OptionalBridgedBasicBlock) align 1 %1) #0
```

Notably `returnInfo.isIndirect` is false, `getSILFuncConventions().getNumIndirectSILResults()` is 0, `returnInfo.getKind()` is `ABIArgInfo::Ignore`, (and `returnInfo.getInReg()` is false or would cause an assert failure.)

Here's part of the debug print code I added:

In `expandExternalSignatureTypes`
```
    llvm::dbgs() << "expandExternalSignatureTypes on\n";
    FnType->dump();
    llvm::dbgs() << "FnType->getRepresentation() " << (uint32_t)(FnType->getRepresentation()) << "\n";
    llvm::dbgs() << "SILResultTy ";
    SILResultTy.dump();
    llvm::dbgs() << "clangResultTy ";
    clangResultTy.dump();
    llvm::dbgs() << "returnInfo.isIndirect " << returnInfo.isIndirect() << "\n";
    llvm::dbgs() << "getSILFuncConventions().getNumIndirectSILResults() "
                 << getSILFuncConventions().getNumIndirectSILResults() << "\n";
    llvm::dbgs() << "getSILFuncConventions().getNumIndirectSILResults() > 0 "
                 << (getSILFuncConventions().getNumIndirectSILResults() > 0) << "\n";
    llvm::dbgs() << "returnInfo.getKind() " << (uint32_t)(returnInfo.getKind()) << "\n";
    llvm::dbgs() << "returnInfo.isInAlloca() " << returnInfo.isInAlloca() << "\n";
    if (returnInfo.isInAlloca()) {
      llvm::dbgs() << "returnInfo.getInAllocaSRet() "
                   << returnInfo.getInAllocaSRet()
                   << "\n";
      llvm::dbgs() << "returnInfo.getInAllocaIndirect() "
                   << returnInfo.getInAllocaIndirect()
                   << "\n";
    }
    if (returnInfo.isDirect() || returnInfo.isExtend() || returnInfo.isIndirect())
      llvm::dbgs() << "returnInfo.getInReg() " << returnInfo.getInReg() << "\n";
```

In `CreateCallOrInvoke`
```
        if (func->hasParamAttribute(argIndex, llvm::Attribute::InReg) &&
            !attrs.hasParamAttr(argIndex, llvm::Attribute::InReg)) {
          llvm::dbgs() << "CreateCallOrInvoke: found a missing inreg\n";
          func->dump();
        }
```

I see that in `SyncCallEmission::setARgs`, `getNumIndirectSILResults` in `passIndirectResults` is returning 1 here, as opposed to 0 in `expandExternalSignatureTypes`. So something may be different there, but I have yet to understand that code.

hjyamauchi commented:
Ah. Interesting. I am going to experiment with `arrangeCXXMethodType`

hjyamauchi commented:
I used `arrangeCXXMethodCall` and it appears to be working now! The `returnInfo` says it's indirect and `sret`/`inreg` are added in `expandExternalSignatureTypes`.

```
expandExternalSignatureTypes on
(sil_function_type type="@convention(cxx_method) (@in_guaranteed BridgedFunction) -> OptionalBridgedBasicBlock"
  (input=struct_type decl="InRegSRet.(file).BridgedFunction")
  (result=struct_type decl="InRegSRet.(file).OptionalBridgedBasicBlock")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
FnType->getRepresentation() 12
SILResultTy $OptionalBridgedBasicBlock
clangResultTy RecordType 0x81031d0 'struct OptionalBridgedBasicBlock' imported
`-CXXRecord 0xfe48a8 'OptionalBridgedBasicBlock'
returnInfo.isIndirect 1
getSILFuncConventions().getNumIndirectSILResults() 0
getSILFuncConventions().getNumIndirectSILResults() > 0 0
returnInfo.getKind() 2
returnInfo.isInAlloca() 0
returnInfo.getInReg() 1
addIndirectResultAttributes : added sret
addIndirectResultAttributes : added inreg
```

The diff currently looks like the following:

In `expandExternalSignatureTypes`
```
-  auto &FI = clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
-                                             clangResultTy, paramTys, extInfo,
-                                             clang::CodeGen::RequiredArgs::All);
+  bool isCXXMethod =
+      FnType->getRepresentation() == SILFunctionTypeRepresentation::CXXMethod;
+  auto &FI = isCXXMethod ?
+      clang::CodeGen::arrangeCXXMethodCall(IGM.ClangCodeGen->CGM(),
+                                           clangResultTy, paramTys, extInfo,
+                                           clang::CodeGen::RequiredArgs::All) :
+      clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
+                                              clangResultTy, paramTys, extInfo,
+                                              clang::CodeGen::RequiredArgs::All);
 ```

In the `llvm-project` repo
```
diff --git a/clang/include/clang/CodeGen/CodeGenABITypes.h b/clang/include/clang/CodeGen/CodeGenABITypes.h
index 0ba6c2019285..0d38766cac34 100644
--- a/clang/include/clang/CodeGen/CodeGenABITypes.h
+++ b/clang/include/clang/CodeGen/CodeGenABITypes.h
@@ -75,6 +75,12 @@ const CGFunctionInfo &arrangeCXXMethodType(CodeGenModule &CGM,
                                            const FunctionProtoType *FTP,
                                            const CXXMethodDecl *MD);

+const CGFunctionInfo &arrangeCXXMethodCall(CodeGenModule &CGM,
+                                           CanQualType returnType,
+                                           ArrayRef<CanQualType> argTypes,
+                                           FunctionType::ExtInfo info,
+                                           RequiredArgs args);
+
 const CGFunctionInfo &arrangeFreeFunctionCall(CodeGenModule &CGM,
                                               CanQualType returnType,
                                               ArrayRef<CanQualType> argTypes,
diff --git a/clang/lib/CodeGen/CodeGenABITypes.cpp b/clang/lib/CodeGen/CodeGenABITypes.cpp
index a6073e1188d6..283aa8425867 100644
--- a/clang/lib/CodeGen/CodeGenABITypes.cpp
+++ b/clang/lib/CodeGen/CodeGenABITypes.cpp
@@ -59,6 +59,17 @@ CodeGen::arrangeCXXMethodType(CodeGenModule &CGM,
   return CGM.getTypes().arrangeCXXMethodType(RD, FTP, MD);
 }

+const CGFunctionInfo &
+CodeGen::arrangeCXXMethodCall(CodeGenModule &CGM,
+                              CanQualType returnType,
+                              ArrayRef<CanQualType> argTypes,
+                              FunctionType::ExtInfo info,
+                              RequiredArgs args) {
+  return CGM.getTypes().arrangeLLVMFunctionInfo(returnType,
+                                                FnInfoOpts::IsInstanceMethod,
+                                                argTypes, info, {}, args);
+}
+
 const CGFunctionInfo &
 CodeGen::arrangeFreeFunctionCall(CodeGenModule &CGM,
                                  CanQualType returnType,

```

I added `arrangeCXXMethodCall` in `CodeGenABITypes.h` instead of using the existing `arrangeCXXMethodType` to keep the same set of arguments as `arrangeFreeFunctionCall`.

Is there a reason to not to add `arrangeCXXMethodCall` and to use `arrangeCXXMethodType`? If so, how? @rjmccall 

I also need to test the whole win/arm64 toolchain bootstrap works with this.

rjmccall commented:
> I added `arrangeCXXMethodCall` in `CodeGenABITypes.h` instead of using the existing `arrangeCXXMethodType` to keep the same set of arguments as `arrangeFreeFunctionCall`.
> 
> Is there a reason to not to add `arrangeCXXMethodCall` and to use `arrangeCXXMethodType`? If so, how? @rjmccall

No, it's abstractly the right thing to do.  It shouldn't matter practically because IIRC declaration vs. type arrangement only matters for variadic calls, which Swift currently never emits, but I do support doing it.  Please upstream that patch to LLVM if you don't mind.

hjyamauchi commented:
Yeah, will upstream the patch to LLVM. ty!

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1462,9 +1462,13 @@ void SignatureExpansion::expandExternalSignatureTypes() {
+      clang::CodeGen::arrangeCXXMethodCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, clang::CodeGen::RequiredArgs::All) :
+      clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, clang::CodeGen::RequiredArgs::All);
```
Now that we're arranging the call properly, can we change the places in IRGen that reorder the `sret` argument on Windows targets to just use `isSRetAfterThis()`?  This probably wasn't being set before, so it's understandable that we had to adjust it, but now that we've got an accurate `CGFunctionInfo`, we should be able to just trust this bit.

rjmccall reviewed (commented):
Thanks, this looks great!  One last substantive request, and then otherwise this is approved as far as I'm concerned.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
An update:

I merged the spun-off Clang PR: https://github.com/llvm/llvm-project/pull/111597. Thanks for reviewing, John. I'm working on cherrypicking it into swiftlang.

Re: the above refactoring to do with `isSRetAfterThis`, I attempted this refactoring and it worked up to some extent but I hit a few cases in the swift tests where the use of `isSRetAfterThis` didn't preserve the existing behaviors.

I changed the four code locations in `GenCalls.cpp` (I attach the patch/diff at the bottom.) and replaced the existing conditions that check for the windows build and the C++ method, etc. with the new ones that use `isSRetAfterThis` and also inserted asserts that check the equivalence between the old and the new conditions. But I encountered a few assert failure cases in particular with some C++ `operator+=`'s. It seems like they use an `sret` argument but their `CGFunctionInfo` don't indicate that they are indirect-return (i.e., `isIndirect()==false`). (Note `isSRetAfterThis` can be called only when `isIndirect()==true`.) @rjmccall Do you see what I'm missing here? 

The asserts failed in the following two tests along with some debug prints:

Test 1: `Interop/Cxx/class/method/msvc-abi-return-indirect-trivial-record.swift`
```
llvm::FunctionType
void (ptr, ptr, i64)
OrigFunctionType
(sil_function_type type="@convention(cxx_method) (LoadableIntWrapper, @inout LoadableIntWrapper) -> ()"
  (input=struct_type decl="MsvcUseVecIt.(file).LoadableIntWrapper")
  (input=struct_type decl="MsvcUseVecIt.(file).LoadableIntWrapper")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
SubstFunctionType
(sil_function_type type="@convention(cxx_method) (LoadableIntWrapper, @inout LoadableIntWrapper) -> ()"
  (input=struct_type decl="MsvcUseVecIt.(file).LoadableIntWrapper")
  (input=struct_type decl="MsvcUseVecIt.(file).LoadableIntWrapper")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
IGF.IGM.Triple.isWindowsMSVCEnvironment() 1
getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod 1
substResultType.isVoid() 1
getCallee().getFunctionPointer().getRawPointer() 1
FI != nullptr 1
FI->getReturnInfo().isIndirect() 0
FI->getReturnInfo().getKind() 4
Assertion failed: (IGF.IGM.Triple.isWindowsMSVCEnvironment() && getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod && substResultType.isVoid() && getCallee().getFunctionPointer().getRawPointer()) == (FI && FI->getReturnInfo().isIndirect() && FI->getReturnInfo().isSRetAfterThis() && substResultType.isVoid() && getCallee().getFunctionPointer().getRawPointer()), file S:\SourceCache\swift\lib\IRGen\GenCall.cpp, line 3623
```

The `operator+=` in source code.
```
struct LoadableIntWrapper {
    int value;

    LoadableIntWrapper operator+=(LoadableIntWrapper rhs) {
        value += rhs.value;
        return *this;
    }
};
```

Test 2: `Interop/Cxx/operators/member-inline.swift`
```
llvm::FunctionType
void (ptr, ptr, i64)
OrigFunctionType
(sil_function_type type="@convention(cxx_method) (LoadableIntWrapper, @inout DerivedFromLoadableIntWrapperWithUsingDecl) -> ()"
  (input=struct_type decl="MemberInline.(file).LoadableIntWrapper")
  (input=struct_type decl="MemberInline.(file).DerivedFromLoadableIntWrapperWithUsingDecl")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
SubstFunctionType
(sil_function_type type="@convention(cxx_method) (LoadableIntWrapper, @inout DerivedFromLoadableIntWrapperWithUsingDecl) -> ()"
  (input=struct_type decl="MemberInline.(file).LoadableIntWrapper")
  (input=struct_type decl="MemberInline.(file).DerivedFromLoadableIntWrapperWithUsingDecl")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
IGF.IGM.Triple.isWindowsMSVCEnvironment() 1
getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod 1
substResultType.isVoid() 1
getCallee().getFunctionPointer().getRawPointer() 1
FI != nullptr 1
FI->getReturnInfo().isIndirect() 0
FI->getReturnInfo().getKind() 4
Assertion failed: (IGF.IGM.Triple.isWindowsMSVCEnvironment() && getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod && substResultType.isVoid() && getCallee().getFunctionPointer().getRawPointer()) == (FI && FI->getReturnInfo().isIndirect() && FI->getReturnInfo().isSRetAfterThis() && substResultType.isVoid() && getCallee().getFunctionPointer().getRawPointer()), file S:\SourceCache\swift\lib\IRGen\GenCall.cpp, line 3623
```

The `operator+=` in source code.
```
struct LoadableIntWrapper {
  int value;
  LoadableIntWrapper operator-(LoadableIntWrapper rhs) {
    return LoadableIntWrapper{.value = value - rhs.value};
  }

  LoadableIntWrapper operator+=(LoadableIntWrapper rhs) {
    value += rhs.value;
    return *this;
  }
  ...
};
struct DerivedFromLoadableIntWrapperWithUsingDecl : private LoadableIntWrapper {
  using LoadableIntWrapper::operator-;
  using LoadableIntWrapper::operator+=;
 ...
};
```


The patch/diff
The four code locations are marked with: `*** (1) ***`, `*** (2) ***`, etc. The above two failures happen in `*** (4) ***`.
```
diff --git a/lib/IRGen/Callee.h b/lib/IRGen/Callee.h
index 66199d8819f..6232d9436f8 100644
--- a/lib/IRGen/Callee.h
+++ b/lib/IRGen/Callee.h
@@ -563,7 +563,7 @@ namespace irgen {
 
     const FunctionPointer &getFunctionPointer() const { return Fn; }
 
-    llvm::FunctionType *getLLVMFunctionType() {
+    llvm::FunctionType *getLLVMFunctionType() const {
       return Fn.getFunctionType();
     }
 
diff --git a/lib/IRGen/GenCall.cpp b/lib/IRGen/GenCall.cpp
index 660e8d22f9f..39ef485b27b 100644
--- a/lib/IRGen/GenCall.cpp
+++ b/lib/IRGen/GenCall.cpp
@@ -1462,9 +1462,15 @@ void SignatureExpansion::expandExternalSignatureTypes() {
   // Generate function info for this signature.
   auto extInfo = clang::FunctionType::ExtInfo();
 
-  auto &FI = clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
-                                             clangResultTy, paramTys, extInfo,
-                                             clang::CodeGen::RequiredArgs::All);
+  bool isCXXMethod =
+      FnType->getRepresentation() == SILFunctionTypeRepresentation::CXXMethod;
+  auto &FI = isCXXMethod ?
+      clang::CodeGen::arrangeCXXMethodCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, {},
+          clang::CodeGen::RequiredArgs::All) :
+      clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, {},
+          clang::CodeGen::RequiredArgs::All);
   ForeignInfo.ClangInfo = &FI;
 
   assert(FI.arg_size() == paramTys.size() &&
@@ -1586,9 +1592,11 @@ void SignatureExpansion::expandExternalSignatureTypes() {
   if (returnInfo.isIndirect()) {
     auto resultType = getSILFuncConventions().getSingleSILResultType(
         IGM.getMaximalTypeExpansionContext());
-    if (IGM.Triple.isWindowsMSVCEnvironment() &&
-        FnType->getRepresentation() ==
-            SILFunctionTypeRepresentation::CXXMethod) {
+    // *** (1) *** This passes tests.
+    assert((IGM.Triple.isWindowsMSVCEnvironment() &&
+            FnType->getRepresentation() == SILFunctionTypeRepresentation::CXXMethod) ==
+           returnInfo.isSRetAfterThis());
+    if (returnInfo.isSRetAfterThis()) {
       // Windows ABI places `this` before the
       // returned indirect values.
       emitArg(0);
@@ -2555,11 +2563,22 @@ public:
 
         // Windows ABI places `this` before the
         // returned indirect values.
-        bool isThisFirst = IGF.IGM.Triple.isWindowsMSVCEnvironment();
-        if (!isThisFirst)
+        // *** (2) *** This passes tests.
+        assert(getCallee().getForeignInfo().ClangInfo != nullptr);
+        auto &returnInfo =
+            getCallee().getForeignInfo().ClangInfo->getReturnInfo();
+        if (fnConv.getNumIndirectSILResults() > 0) {
+          if (!returnInfo.isIndirect()) {
+            llvm::dbgs() << "returnInfo.isIndirect() " << returnInfo.isIndirect() << "\n";
+            llvm::dbgs() << "fnConv.getNumIndirectSILResults() " << fnConv.getNumIndirectSILResults() << "\n";
+          }
+          assert(returnInfo.isIndirect());
+          assert(returnInfo.isSRetAfterThis() == IGF.IGM.Triple.isWindowsMSVCEnvironment());
+        }
+        if (returnInfo.isIndirect() && !returnInfo.isSRetAfterThis())
           passIndirectResults();
         adjusted.add(arg);
-        if (isThisFirst)
+        if (returnInfo.isIndirect() && returnInfo.isSRetAfterThis())
           passIndirectResults();
       }
 
@@ -3148,9 +3167,46 @@ void CallEmission::emitToUnmappedMemory(Address result) {
   assert(LastArgWritten == 1 && "emitting unnaturally to indirect result");
 
   Args[0] = result.getAddress();
-  if (IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
-      getCallee().getRepresentation() ==
-          SILFunctionTypeRepresentation::CXXMethod &&
+
+  // *** (3) *** 
+  auto *FI = getCallee().getForeignInfo().ClangInfo;
+  if ((IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
+       getCallee().getRepresentation() ==
+           SILFunctionTypeRepresentation::CXXMethod &&
+       Args[1] == getCallee().getCXXMethodSelf()) !=
+      (FI && FI->getReturnInfo().isIndirect() &&
+       FI->getReturnInfo().isSRetAfterThis() &&
+       Args[1] == getCallee().getCXXMethodSelf())) {
+    llvm::dbgs() << "IGF.IGM.Triple.isWindowsMSVCEnvironment() "
+                 << IGF.IGM.Triple.isWindowsMSVCEnvironment() << "\n";
+    llvm::dbgs() << "getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod "
+                 << (getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod) << "\n";
+    llvm::dbgs() << "FI != nullptr " << (FI != nullptr) << "\n";
+    if (FI) {
+      llvm::dbgs() << "FI->getReturnInfo().isIndirect() "
+                   << FI->getReturnInfo().isIndirect() << "\n";
+      llvm::dbgs() << "FI->getReturnInfo().getKind() "
+                   << (unsigned)FI->getReturnInfo().getKind() << "\n";
+      if (FI->getReturnInfo().isIndirect()) {
+        llvm::dbgs() << "FI->getReturnInfo().isSRetAfterThis()"
+                     << FI->getReturnInfo().isSRetAfterThis() << "\n";
+      }
+    }
+    llvm::dbgs() << "Args.size() " << Args.size() << "\n";
+    if (Args.size() >= 2) {
+      llvm::dbgs() << "Args[1] == getCallee().getCXXMethodSelf() "
+                   << (Args[1] == getCallee().getCXXMethodSelf()) << "\n";
+    }
+  }
+  assert((IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
+          getCallee().getRepresentation() ==
+              SILFunctionTypeRepresentation::CXXMethod &&
+          Args[1] == getCallee().getCXXMethodSelf()) ==
+         (FI && FI->getReturnInfo().isIndirect() &&
+          FI->getReturnInfo().isSRetAfterThis() &&
+          Args[1] == getCallee().getCXXMethodSelf()));
+  if (FI && FI->getReturnInfo().isIndirect() &&
+      FI->getReturnInfo().isSRetAfterThis() &&
       Args[1] == getCallee().getCXXMethodSelf()) {
     // C++ methods in MSVC ABI pass `this` before the
     // indirectly returned value.
@@ -3515,10 +3571,58 @@ void CallEmission::emitToExplosion(Explosion &out, bool isOutlined) {
       emitToMemory(temp, substResultTI, isOutlined);
       return;
     }
-    if (IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
-        getCallee().getRepresentation() ==
-            SILFunctionTypeRepresentation::CXXMethod &&
-        substResultType.isVoid()) {
+
+    // *** (4) ***
+    // Didn't work due to operator+=() that returns void but CGFunctionInfo doesn't say it's indirect
+    // Tests:
+    // Interop/Cxx/class/method/msvc-abi-return-indirect-trivial-record.swift
+    // Interop/Cxx/operators/member-inline.swift
+    auto *FI = getCallee().getForeignInfo().ClangInfo;
+    if ((IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
+         getCallee().getRepresentation() ==
+             SILFunctionTypeRepresentation::CXXMethod &&
+         substResultType.isVoid() &&
+         getCallee().getFunctionPointer().getRawPointer()) !=
+        (FI && FI->getReturnInfo().isIndirect() &&
+         FI->getReturnInfo().isSRetAfterThis() &&
+         substResultType.isVoid() &&
+         getCallee().getFunctionPointer().getRawPointer())) {
+      llvm::dbgs() << "llvm::FunctionType " << "\n";
+      getCallee().getLLVMFunctionType()->dump();
+      llvm::dbgs() << "OrigFunctionType " << "\n";
+      getCallee().getOrigFunctionType()->dump();
+      llvm::dbgs() << "SubstFunctionType " << "\n";
+      getCallee().getSubstFunctionType()->dump();
+      llvm::dbgs() << "IGF.IGM.Triple.isWindowsMSVCEnvironment() "
+                   << IGF.IGM.Triple.isWindowsMSVCEnvironment() << "\n";
+      llvm::dbgs() << "getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod "
+                   << (getCallee().getRepresentation() == SILFunctionTypeRepresentation::CXXMethod) << "\n";
+      llvm::dbgs() << "substResultType.isVoid() " << substResultType.isVoid() << "\n";
+      llvm::dbgs() << "getCallee().getFunctionPointer().getRawPointer() "
+                   << (getCallee().getFunctionPointer().getRawPointer() != 0) << "\n";
+      llvm::dbgs() << "FI != nullptr " << (FI != nullptr) << "\n";
+      if (FI) {
+        llvm::dbgs() << "FI->getReturnInfo().isIndirect() "
+                     << FI->getReturnInfo().isIndirect() << "\n";
+        llvm::dbgs() << "FI->getReturnInfo().getKind() "
+                     << (unsigned)FI->getReturnInfo().getKind() << "\n";
+        if (FI->getReturnInfo().isIndirect()) {
+          llvm::dbgs() << "FI->getReturnInfo().isSRetAfterThis() "
+                       << FI->getReturnInfo().isSRetAfterThis() << "\n";
+        }
+      }
+    }
+    assert((IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
+            getCallee().getRepresentation() ==
+                SILFunctionTypeRepresentation::CXXMethod &&
+            substResultType.isVoid() &&
+            getCallee().getFunctionPointer().getRawPointer()) ==
+           (FI && FI->getReturnInfo().isIndirect() &&
+            FI->getReturnInfo().isSRetAfterThis() &&
+            substResultType.isVoid() &&
+            getCallee().getFunctionPointer().getRawPointer()));
+    if (FI && FI->getReturnInfo().isIndirect() &&
+        FI->getReturnInfo().isSRetAfterThis() && substResultType.isVoid()) {
       // Some C++ methods return a value but are imported as
       // returning `Void` (e.g. `operator +=`). In this case
       // we should allocate the correct temp indirect return
```

rjmccall commented:
That seems like a larger problem — that function obviously should not be given a `()` return type in SIL when it clearly returns a non-trivial value.  We could fix that in SIL by "bridging" non-`void` return values into `()` (by ignoring them), but in this case, I feel like we should just be importing the function as having a real return value.  Since it's a (compound) assignment operator, it seems conventionally correct to infer `@discardableResult`, though.

hjyamauchi commented:
@rjmccall

`CGFunctionInfo` is from the Clang code. So, that's just how Clang handles `operator+=` and not a bug? If so, I'm curious how this isn't a problem in Clang, which may need to do similar `isIndirect` and `isSRetAfterThis` checks.

Is "importing the function as having a real return value" referring to happening in Clang or Swift? If Swift, it isn't a right assumption that we can solely rely on `CGFunctionInfo` to detect an indirect return?

CC @eeckstein @compnerd @ahatanaka @hyp Any thoughts?

hjyamauchi commented:
@swift-ci please test

rjmccall commented:
> @rjmccall
> 
> `CGFunctionInfo` is from the Clang code. So, that's just how Clang handles `operator+=` and not a bug? If so, I'm curious how this isn't a problem in Clang, which may need to do similar `isIndirect` and `isSRetAfterThis` checks.

It's not an issue in Clang because Clang is emitting the call using the actual signature of the function.  Swift could do the same if it propagated the Clang function type in the `SILFunctionType`.  Because it doesn't do that, Swift has to regenerate the Clang function type from the components of the `SILFunctionType`, which is why it's a problem that the `SILFunctionType` is incorrectly saying that the function doesn't have a return value.

We're importing this operator at the AST level as if it returned `void`; that's then just carrying through into the SIL type.  I assume the importer is doing a blanket suppression of the return value of assignment operators.  @egorzhdan probably knows why, but I imagine it's because assignment operators usually return references and people usually want to ignore them.  That's definitely biting us here, though.  It's okay on most platforms to call a pointer-returning function as a `void`-returning function, but that's not a totally portable guarantee (e.g. I imagine it blows up on wasm), and it's definitely not true for arbitrary return types — for example, on i386, `float` functions return on the x87 stack, and calling them as `void` functions will leave the stack unbalanced. In this case, we're both missing an `sret` and potentially leaking a value (if the return type were non-trivial).  As I said before, the right fix is probably just to import compound-assignment operators with their actual return type, maybe decorated as `@discardableResult` for usability reasons.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@rjmccall Okay. Would you elaborate on "import compound-assignment operators with their actual return type, maybe decorated as @discardableResult for usability reasons" (like what code to modify, etc. if you are familiar)? I'll see what I can do if you can provide guidance. @egorzhdan I'd appreciate your thoughts.

rjmccall commented:
> @rjmccall Okay. Would you elaborate on "import compound-assignment operators with their actual return type, maybe decorated as @discardableResult for usability reasons" (like what code to modify, etc. if you are familiar)? I'll see what I can do if you can provide guidance. @egorzhdan I'd appreciate your thoughts.

Looks like this is the line:

```
https://github.com/swiftlang/swift/blob/940730efb66c22439b32feaf585ba5348fcff907/lib/ClangImporter/ImportType.cpp#L2205
```

I'd say that this should check that the result is a reference (and maybe a reference specifically to the `this` type?) before dropping it.  I believe we automatically add `@discardableResult` to imported declarations with non-`Void` return types, so that much is handled already.

hjyamauchi commented:
As suggested, dropping the return type from the `operator+=` etc. only when the return type is of a reference type seems to make [the above assertion failures](https://github.com/swiftlang/swift/pull/76324#issuecomment-2420145075) go away. I also tried to limit this to the this pointer case but it seems to cause quite a few other test failures maybe because the conformance checking relies on the return type being void (the details at the bottom). I think it's sufficient to drop the return type only when the return type is a reference type because that's probably what the original intention of the return-type-dropping code and a reference-type return type means it's unlikely an indirect return, which solves this `CGFunctionInfo`/`operator+=` issue with the above assert failures. Please take another look @rjmccall 

The test failures (with restricting to this pointer)
```
******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/class/protocol-conformance-typechecker.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:48:15: error: unexpected note produced: protocol requires function '+=' with type '(inout HasOperatorPlusEqual<CInt>, Int32) -> ()'
  static func +=(lhs: inout Self, x: Int32)
              ^
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:51:1: error: unexpected error produced: type 'HasOperatorPlusEqual<CInt>' does not conform to protocol 'HasOperatorPlusEqualProtocol'
extension HasOperatorPlusEqualInt : HasOperatorPlusEqualProtocol {}
^
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:51:1: error: unexpected note produced: add stubs for conformance
extension HasOperatorPlusEqualInt : HasOperatorPlusEqualProtocol {}
^

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/msvcprt-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\msvcprt-module-interface.swift:15:18: error: CHECK-STRING: expected string not found in input
// CHECK-STRING: struct basic_string<CChar, char_traits<CChar>, allocator<CChar>> : CxxRandomAccessCollection {
                 ^
<stdin>:23948:54: note: scanning from here
 static func to_wstring(_ _Val: Int32) -> std.wstring
                                                     ^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\msvcprt-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-collection-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection-module-interface.swift:17:11: error: CHECK: expected string not found in input
// CHECK: struct SimpleCollectionReadOnly : CxxRandomAccessCollection {
          ^
<stdin>:167:2: note: scanning from here
}
 ^
<stdin>:168:1: note: possible intended match here
struct SimpleCollectionReadOnly : CxxConvertibleToCollection {
^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-collection.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection.swift:32:17: error: value of type 'SimpleCollectionReadOnly' has no member 'first'
30 | CxxCollectionTestSuite.test("SimpleCollectionReadOnly as Swift.Collection") {
31 |   let c = SimpleCollectionReadOnly()
32 |   expectEqual(c.first, 1)
   |                 `- error: value of type 'SimpleCollectionReadOnly' has no member 'first'
33 |   expectEqual(c.last, 5)
34 |

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-iterator-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-iterator-module-interface.swift:22:11: error: CHECK: expected string not found in input
// CHECK: struct ConstRACIteratorRefPlusEq : UnsafeCxxRandomAccessIterator, UnsafeCxxInputIterator {
          ^
<stdin>:70:2: note: scanning from here
}
 ^
<stdin>:72:1: note: possible intended match here
struct ConstRACIteratorRefPlusEq : UnsafeCxxInputIterator {
^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-iterator-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-span.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-span.swift:185:18: error: cannot use mutating getter on immutable value: 's' is a 'let' constant
180 |   var arr: [Int32] = [1, 2, 3]
181 |   arr.withUnsafeMutableBufferPointer{ ubpointer in
182 |     let s = initSpan(ubpointer.baseAddress!, ubpointer.count)
    |     `- note: change 'let' to 'var' to make it mutable
183 |     expectEqual(s.size(), 3)
184 |     expectFalse(s.empty())
185 |     expectEqual(s[0], 1)
    |                  `- error: cannot use mutating getter on immutable value: 's' is a 'let' constant
186 |     expectEqual(s[1], 2)
187 |     expectEqual(s[2], 3)

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-string.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-string.swift:372:14: error: for-in loop requires 'std.string' (aka 'std.basic_string<CChar, char_traits<CChar>, allocator<CChar>>') to conform to 'Sequence'
370 |     let cxx1 = std.string()
371 |     var iterated = false
372 |     for _ in cxx1 {
    |              `- error: for-in loop requires 'std.string' (aka 'std.basic_string<CChar, char_traits<CChar>, allocator<CChar>>') to conform to 'Sequence'
373 |         iterated = true
374 |     }

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-vector.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:32:20: error: cannot convert value of type '[Any]' to expected argument type 'std.allocator<CInt>'
 30 |
 31 | StdVectorTestSuite.test("VectorOfInt.init(sequence)") {
 32 |     let v = Vector([])
    |                    `- error: cannot convert value of type '[Any]' to expected argument type 'std.allocator<CInt>'
 33 |     expectEqual(v.size(), 0)
 34 |     expectTrue(v.empty())

S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:155:15: error: value of type 'Vector' (aka 'std.vector<CInt, allocator<CInt>>') has no member 'map'
153 |     fill(vector: &v)
154 |
155 |     let a = v.map { $0 + 5 }
    |               `- error: value of type 'Vector' (aka 'std.vector<CInt, allocator<CInt>>') has no member 'map'
156 |     expectEqual(a, [6, 7, 8])
157 | }

S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:174:14: error: for-in loop requires 'VectorSubclass' to conform to 'Sequence'
172 |
173 |     var count: CInt = 1
174 |     for e in v {
    |              `- error: for-in loop requires 'VectorSubclass' to conform to 'Sequence'
175 |         expectEqual(e, count)
176 |         count += 1

******************** TEST 'Swift(windows-x86_64) :: Interop/CxxToSwiftToCxx/span/span-execution.cpp' FAILED ********************
S:\b\5\tools\swift\test-windows-x86_64\Interop\CxxToSwiftToCxx\span\Output\span-execution.cpp.tmp/use-span.swift:12:15: error: cannot use mutating getter on immutable value: 'sp' is a 'let' constant
10 |
11 | public func printSpan(_ sp: Span) {
12 |   print("{\(sp[0]), \(sp[1]), \(sp[2])}")
   |               `- error: cannot use mutating getter on immutable value: 'sp' is a 'let' constant
13 | }
14 |

S:\b\5\tools\swift\test-windows-x86_64\Interop\CxxToSwiftToCxx\span\Output\span-execution.cpp.tmp/use-span.swift:28:19: error: value of type 'Span' (aka 'std.span<CInt, _CUnsignedLongLong_-1>') has no member 'map'
26 |
27 | public func mapSpan(_ sp: Span) {
28 |   let result = sp.map { $0 + 3 }
   |                   `- error: value of type 'Span' (aka 'std.span<CInt, _CUnsignedLongLong_-1>') has no member 'map'
29 |   print(result)
30 | }
```

rjmccall reviewed (approved):
LGTM

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@swift-ci please test

rjmccall reviewed (approved):
Ah, I thought it looked a little short. LGTM.

hjyamauchi commented:
Thanks, John! I will look into the test failures.

hjyamauchi committed:
Fix the IR gen for C++ method calls and refactor around CGFunctionInfo

In GenCall, fix the IR gen for C++ method calls as under MSVC as the
calling conventions for free functions and C++ methods can be
different. This also fixes the missing inreg (on sret arguments)
issues on Windows ARM64. Also refactor to use CGFunctionInfo
returnInfo isSretAfterThis to detect when to reorder the sret and the
this arguments under MSVC.

In ClagImporter, don't drop the return type for the compound
assignment operators such as operator+= when the return value is a
reference so that the CGFunctionInfo will be correctly indicate an
indirect return for the compound assignment operators.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@swift-ci please test Windows Platform

hjyamauchi commented:
@swift-ci please test Windows Platform

hjyamauchi commented:
@swift-ci please test Windows Platform

hjyamauchi commented:
@swift-ci please test Windows Platform

hjyamauchi commented:
@rjmccall Can you take a quick look after the test fix? (for the return type change, additional attributes like `noalias`, `nocapture` and a C++ struct type name change as a consequence of the code gen change.) The CI all passed.

rjmccall reviewed (approved):
The test changes look fine, although I'm confused about why we ever used 16-byte alignment for this type.  I [verified](https://godbolt.org/z/a9r4e33Kh) that MSVC only guarantees ordinary ABI alignment for sret results, as you'd normally expect.

hjyamauchi commented:
Thanks, John! I am not sure why the 16-byte alignment there, either 🤔 but I agree with you. If I have a chance, I'll think about this. I'll merge this, and move onto cherrypicking this along with the other commits to 6 and 5.10 to make progress in resurrecting the windows arm64 compiler.
